hilti/toolchain/include/compiler/detail/cfg.h

```diff
@@ -0,0 +1,142 @@
+};
+
+// A meta node for the start of a control flow.
+struct Start : MetaNode {
+    Start() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(Start, MetaNode, final);
+};
+
+// A meta node for the end of a control flow.
+struct End : MetaNode {
+    End() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(End, MetaNode, final);
+};
+
+// A meta node joining or splitting control flow with no matching source statement.
+struct Flow : MetaNode {
+    Flow() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(Flow, MetaNode, final);
+};
+
+// A meta node to signify end of a scope carrying the source range of that scope.
+struct ScopeEnd : MetaNode {
+    ScopeEnd(const Node* scope) : MetaNode(NodeTags), scope(scope) {
+        assert(scope); // Should always contain a valid scope.
+    }
+
+    HILTI_NODE_1(ScopeEnd, MetaNode, final);
+
+    const Node* scope;
+};
+
+struct Reachability {
+    std::unordered_set<Node*> in;
+    std::unordered_set<Node*> out;
+};
+
+struct Transfer {
+    std::unordered_set<Node*> use;
+    std::unordered_map<Node*, Node*> gen;
+    std::unordered_map<Node*, std::unordered_set<Node*>> kill;
+    std::unordered_set<Node*> aliases;
+
+    bool keep = false; // Whether this node should be kept.
+
+    std::optional<Reachability> reachability;
+};
+
+class CFG {
+public:
+    CFG(const Node* root);
+
+    template<typename T, typename... Args, typename = std::enable_if_t<std::is_base_of_v<MetaNode, T>>>
+    Node* create_meta_node(Args... args) {
+        auto n = std::make_unique<T>(args...);
+        auto* r = n.get();
+        _meta_nodes.insert(std::move(n));
+        return r;
+    }
+
+    Node* getOrAddNode(Node* n);
+    void addEdge(Node* from, Node* to);
+
+    // Add flow for globals if `root` corresponds to a global module block.
+    [[nodiscard]] Node* addGlobals(Node* parent, const Node& root);
+
+    [[nodiscard]]
+    Node* addBlock(Node* parent, const Nodes& stmts, const Node* scope);
+    [[nodiscard]]
+    Node* addFor(Node* parent, const statement::For& for_);
+    [[nodiscard]]
+    Node* addWhile(Node* parent, const statement::While& while_, Node* scope_end);
+    [[nodiscard]]
+    Node* addIf(Node* parent, const statement::If& if_);
+    [[nodiscard]]
+    Node* addTryCatch(Node* parent, const statement::Try& try_);
+    [[nodiscard]]
+    Node* addReturn(Node* parent, const Node* expression);
+    [[nodiscard]]
+    Node* addThrow(Node* parent, statement::Throw& throw_, Node* scope_end);
+    [[nodiscard]]
+    Node* addCall(Node* parent, operator_::function::Call* call);
+
+    std::unordered_set<Node*> unreachableNodes() const;
+
+    std::string dot() const;
+
+    void populateDataflow();
+    void populateReachableExpressions();
+    std::vector<Node*> unreachableStatements() const;
+
+    util::graph::DirectedGraph<Node*, uintptr_t> g;
+
+private:
+    std::unordered_set<std::unique_ptr<MetaNode>> _meta_nodes;
+    std::unordered_map<Node*, Transfer> _dataflow;
+    Node* _begin;
+    Node* _end;
```

Comment:
Is `_end` the end of the *block*, the *function*, etc?

More of asking as a way to say that something is slightly broken, but may not currently manifest much. Given this code:

```
module Test {

import hilti;

function void f() {
    {
        hilti::print("hey");
        return;
    }
    hilti::print("bad!");
}

f();

}
```

The optimization pass doesn't remove `hilti::print` with the scope, but does without it. `return` should point to the end of the *function*, not block